nixos/modules/services/databases/lldap.nix

```diff
@@ -102,12 +102,66 @@ in
             default = "sqlite://./users.db?mode=rwc";
             example = "postgres://postgres-user:password@postgres-server/my-database";
           };
+
+          ldap_user_pass_file = mkOption {
+            type = types.nullOr types.str;
+            default = null;
+            description = ''
+              Path to a file containing the default admin password.
+            '';
+          };
+
+          force_ldap_user_pass_reset = mkOption {
+            type = types.oneOf [
+              types.bool
+              (types.enum [ "always" ])
+            ];
+            default = false;
+            description = ''
+              Force reset of the admin password.
+
+              Break glass in case of emergency: if you lost the admin password, you
+              can set this to true to force a reset of the admin password to the value
+              of `ldap_user_pass_file`.
+
+              Alternatively, you can set it to `"always"` to reset every time the server starts
+              which makes for a more declarative configuration.
+
+              The difference between `true` and `"always"` is the former is intended for a one time fix
+              while the latter is intended for a declarative workflow. In practice, the result
+              is the same: the password gets reset. The only practical difference is the former
+              outputs a warning message while the latter outputs an info message.
+            '';
+          };
+
+          jwt_secret_file = mkOption {
+            type = types.nullOr types.str;
+            default = null;
+            description = ''
+              Path to a file containing the JWT secret.
+            '';
+          };
         };
       };
+
+      # TOML does not allow null values, so we use null to omit those fields
+      apply = lib.filterAttrsRecursive (_: v: v != null);
     };
   };
 
   config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion =
+          cfg.settings.ldap_user_pass_file == "" || cfg.settings.force_ldap_user_pass_reset != false;
+        message = ''
+          The default admin password is set declaratively with `ldap_user_pass_file` option but the `force_ldap_user_pass_reset` is set to `false`.
+          This means the admin password can be changed through the UI and will drift from the one defined in your nix config.
+          Please set the `force_ldap_user_pass_reset` option to `true` or `"always"`.
+        '';
+      }
+    ];
```

Comment:
Wouldn't you know it, I ran into this exact assertion, when trying the patch ...

... so, thinking about it, I do appreciate the heads-up about this, and especially if one keeps ssh public keys in ldap (does lldap support that yet?), it could be critical to maintain system access. I'll probably change that in my config.

However, I'm not convinced, that there is absolutely no use case ever, for running with a password file, but no force reset (which is what we're saying with this assertion). After all, it's the default behavior in lldap, and if it really was such a no-go, then lldap should probably be changed to not allow it, or at least not default to it.

So I'd tend towards thinking that this assertion is overreach and should be downgraded to a warning.

What do you think?

Reply:
That makes total sense to me. The "next" PR https://github.com/NixOS/nixpkgs/pull/425923 which would be my final goal for LLDAP has it as a [warning](https://github.com/NixOS/nixpkgs/pull/425923/files#diff-e31cdba3cdd927f952d0783bc989ff1e86eb2eb35cc13f79469593a46a21cc70R425) or assertion depending on some options.
I'll switch it to a warning.
Maybe we should even have another option to avoid showing the warning? I've seen that in some modules.

Reply:
> Maybe we should even have another option to avoid showing the warning? I've seen that in some modules. Like a "okay I know what I'm doing don't show the warning anymore" kind of option.

sounds good to me. I was considering introducing `"once"` in addition to `"always"`, but we probably shouldn't add stuff to the settings, that aren't there in upstream.
but a way to suppress an accepted warning would be good. is there a generalized way to do this, perhaps?

Reply:
I'm not sure but I'll see if that exists.